openmc/model/triso.py

```diff
@@ -0,0 +1,172 @@
+            c = openmc.Cell(fill=m)
+            if i == 0:
+                c.region = -surfaces[i]
+            else:
+                c.region = +surfaces[i-1] & -surfaces[i]
+            cells.append(c)
+        self._cells = cells
+        self._surfaces = surfaces
+        self.center = np.asarray(center)
+
+    @property
+    def bounding_box(self):
+        return self.cells[-1].region.bounding_box
+
+    @property
+    def cells(self):
+        return self._cells
+
+    @property
+    def center(self):
+        return self._center
+
+    @property
+    def outside(self):
+        return ~self.cells[-1].region.nodes[-1]
+
+    @center.setter
+    def center(self, center):
+        cv.check_type('TRISO center', center, Iterable, Real)
+        for s in self._surfaces:
+            s.x0, s.y0, s.z0 = center
+        self._center = center
+
+    def classify(self, lattice):
+        """Determine lattice element indices which might contain the TRISO particle.
+
+        Parameters
+        ----------
+        lattice : openmc.RectLattice
+            Lattice to check
+
+        Returns
+        -------
+        list of tuple
+            (z,y,x) lattice element indices which might contain the TRISO
+            particle.
+
+        """
+
+        ll, ur = self.bounding_box
+        if lattice.ndim == 2:
+            (i_min, j_min), p = lattice.find_element(ll)
+            (i_max, j_max), p = lattice.find_element(ur)
+            return list(np.broadcast(*np.ogrid[
+                j_min:j_max+1, i_min:i_max+1]))
+        else:
+            (i_min, j_min, k_min), p = lattice.find_element(ll)
+            (i_max, j_max, k_max), p = lattice.find_element(ur)
+            return list(np.broadcast(*np.ogrid[
+                k_min:k_max+1, j_min:j_max+1, i_min:i_max+1]))
+
+
+def create_triso_lattice(trisos, lower_left, pitch, shape, background):
+    """Create a lattice containing TRISO particles for optimized tracking.
+
+    Parameters
+    ----------
+    trisos : list of openmc.model.TRISO
+        List of TRISO particles to put in lattice
+    lower_left : Iterable of float
+        Lower-left Cartesian coordinates of the lattice
+    pitch : Iterable of float
+        Pitch of the lattice elements in the x-, y-, and z-directions
+    shape : Iterable of float
+        Number of lattice elements in the x-, y-, and z-directions
+    background : openmc.Material
+        A background material that is used anywhere within the lattice but
+        outside a TRISO particle
+
+    Returns
+    -------
+    lattice : openmc.RectLattice
+        A lattice containing the TRISO particles
+
+    """
+
+    lattice = openmc.RectLattice()
+    lattice.lower_left = lower_left
+    lattice.pitch = pitch
+
+    indices = list(np.broadcast(*np.ogrid[:shape[2], :shape[1], :shape[0]]))
+    triso_locations = {idx: [] for idx in indices}
+    for t in trisos:
+        for idx in t.classify(lattice):
+            if idx in sorted(triso_locations):
```

Comment:
To be clear, there should never be a case where
```python
idx in sorted(triso_locations)
```
is False.  It is possible for the `TRISO.classify` method to return indices which do not actually contain the TRISO, but they still need to be valid indices for the lattice.  Otherwise, TRISOs will get cut by the lattice boundary.